### DISCUSSION TITLE
Custom input type registration

### DISCUSSION THREAD

**Discussion by mahmoud** [xstate-wizards/xstate-wizards#83] [Q&A] [answered]
XSW has a great variety of built-in input types, currently:

```
export enum ContentNodeInputType {
  AGE = "age",
  CHECKBOX = "checkbox",
  CURRENCY = "currency",
  DATE = "date",
  INTEGER = "integer",
  NUMBER = "number",
  PASSWORD = "password",
  TEL = "tel",
  TEXT = "text",
}
```

I appreciate the differentiation between e.g., `AGE` and `INTEGER`. Similarly, I have a few input types in mind that _could_ be represented with the built-in types, but I feel would be clearer with their own dedicated presentation.

For instance, if I want to add a `PERCENT` input type, which could present as a 0.0-100.0 slider, I don't see a way to do so without a PR to XSW. Might it be possible to create an API to register that types/components that live outside XSW?

**Comment by youmustfight** [comment, marked as answer]
This exists as is! I guess another thing to do documentation/examples for 😅  With Upsolve, we had two custom components we placed throughout our flows. One we called a `KnowledgeCard` which was a search widget for content/faqs so you can find and read articles inline. On different pages, we'd pass in different args so it preloaded with some articles.

On your `<WizardRunner />` component, you can pass in serializations, which match XState interms of actions/guards, but expand a bit to include validation functions and also components!

```tsx
import { HelpFooter } from "../../..";
import { KnowledgeCard } from "../../..";

export const WizardRunner = (props) => {
  const { data: user } = useLoggedInUser();
  
  return (
    <WizardRunner
      models={upsolveWizardModelsMap}
      spellMap={preparedSpellMap}
      serializations={{
        actions: interviewSerializations.actions,
        components: {
          // Overriding Components (if you want special handling/styling for things like links)
          A: UpLink,
          Button,
          // Custom Components
          KnowledgeCard: KnowledgeCard,
          HelpFooter: HelpFooter,
        },
        functions: interviewSerializations.functions,
        guards: interviewSerializations.guards,
        states: interviewSerializations.states,
        validations: interviewSerializations.validations,
      }}
      // ... 
    />
  );
};
````

Then in your spell, you can reference them like such:

```ts
export const spellMapping = createSpell({
  key: ID_QUESTIONNAIRE_ABOUT_YOU,
  version: "1",
  config: {
    initial: INTERVIEW_INTRO_STATE,
    title: "About You",
  },
  states: {
    introConfirmationAndUnderstanding: {
      content: (ctx) => [
        CONTENT_NODE_BACK,
        { type: "h4", text: "Understanding This Tool" },
        {
          type: "p",
          text: "It’s important that you understand how we can help you and our limitations. Please read the following and confirm that you understand our service by checking these boxes.",
        },
        // ...
        {
          type: "component",
          component: "KnowledgeCard",
          // these attrs are what get passed in as props to your component
          attrs: {
            faq: [
              { label: "How does this tool work?", url: "https://example.org/how-we-work/" },
              { label: "What are our limits?", url: "https://example.org/learn/our-limits" },
            ],
          },
        },
        { type: "component", component: "HelpFooter" },
      ],
      on: {
        SUBMIT: { target: "fillingForms" },
      },
    }
    // ...
})
```

The `KnowledgeCard` component looks like this:
```tsx
export const KnowledgeCard: React.FC<TKnowledgeCardProps> = ({ faq = [] }) => {
    return (
      <StyledKnowledgeCard>
          // ...
      </StyledKnowledgeCard>
    );
}
```

**Reply by youmustfight** [reply]
If you want to do custom inputs that will affect context or send events that you're listening for, components will also have access to the `transition()` function (basically a wrapper for XState `send()`)

```tsx
{
  type: "component",
  component: ({ transition }) => (
    <ButtonLink
      onClick={() => transition({ type: "READ_ARTICLE", data: { key: "articleImprovement" } })}
      href="https://example.org/hc/en-us/articles/360048822654"
      target="_blank"
    >
      Read: "How to tell what will actually improve your situation" [⬈]
    </ButtonLink>
  ),
  alt: "Read article button link",
}
```

I need to refresh myself on this a bit, there's a special `ASSIGN_CONTEXT` even that basically all the inputs are using. I never intend for people to fire that event, since that's what all the inputs are hooked into. I think there may be a better way to extend the input types without having to go all the way to doing the transitions yourself though.

**Comment by mahmoud** [comment]
Ahhhh, I had tried that, but what I missed was that `type: "component"` not `type: "input"`, which just silently renders the fallback input. A warning would be nice for unrecognized `inputType`.

I'm not fully versed in React conventions, but I'd love to see a registration API for this. As is, the built-in input components have special treatment. I'm not yet deep enough to know if that special treatment lets them do anything other components cannot, but IME, having as many builtins as possible go through the same registration route as extensions tends to make for a better extension API. 

At the very least, it would help reduce the size of ContentNode.tsx. And having a registry could allow early detection of references to unrecognized components. I'm trying to take the approach of keeping me spells 100% JSON data to facilitate this end.
